### PR TITLE
Added support for np.ndarrays for SlicedCorpus

### PR DESCRIPTION
**Pull request by mattilyra**
The SlicedCorpus currently only supports slicing using Python lists. It would be quite useful if the slicing could be done similar to NumPy fancy indexing. I think the changes needed are quite simple.

Not sure how to handle the serialisation of the `index` though (turn it into a list - seems silly).

### CODE REVIEW

File: gensim/corpora/indexedcorpus.py [hunk 1]
```diff
@@ -113,7 +115,7 @@ def __getitem__(self, docno):
         if self.index is None:
             raise RuntimeError("cannot call corpus[docid] without an index")
 
-        if isinstance(docno, slice):
+        if isinstance(docno, slice) or isinstance(docno, np.ndarray):
```
Comment (piskvorky) [inline gensim/corpora/indexedcorpus.py]: `isinstance` also accepts multiple class options = tuple as its 2nd parameter.

**Comment by piskvorky** [comment]
Thanks @mattilyra !

Can you also add some tests around this new functionality? These would go into [test/test_corpora.py](https://github.com/piskvorky/gensim/blob/develop/gensim/test/test_corpora.py).

**Comment by mattilyra** [comment]
Sure. I'll add the tests and other missing bits and update the PR once I've got all that done. I was thinking of changing the `index` permanently to a `numpy.array`, are you ok with that? Also are you using `numpy` instead `np` as the NumPy reference? Need to update that as well.

**Comment by piskvorky** [comment]
Great :)

`numpy`/`np`: whatever works for you. Same with `index` as as np array. Just make sure we don't break anything horribly. Also, previously serialized corpora should continue to work (backward compatibility), even if new ones work a bit differently underneath.

**Comment by mattilyra** [comment]
There is one thing I could think of that might break backwards compatibility. If anyone is relying on `IndexedCorpus.index` being a `list` by doing something like `if corpus.index: do_thing()` then that will break.

I thought it's better to break that _always_, instead of changing the `index` on the fly if someone uses fancy indexing. That could lead to extremely weird behaviour, consider the following

    c = IndexedCorpus('dummy.txt') # indexed corpus
    slice_of_c = c[:100:2] # index remains a list
    if slice_of_c.index: # everything is fine because index is a list
        do_something()
    slice_of_c = c[[10, 20, 30, 40]] # index is now a ndarray
    if slice_of_c.index: # FAIL

I think it's better to consistently fail the `if` clauses by always having `index` be `ndarray` instead of it being dependent on runtime state.

File: gensim/corpora/indexedcorpus.py [hunk 2]
```diff
@@ -113,10 +121,12 @@ def __getitem__(self, docno):
         if self.index is None:
             raise RuntimeError("cannot call corpus[docid] without an index")
 
-        if isinstance(docno, slice):
+        if isinstance(docno, (slice, list, numpy.ndarray)):
             return utils.SlicedCorpus(self, docno)
-
-        return self.docbyoffset(self.index[docno])
+        elif isinstance(docno, (int, numpy.int)):
```
Comment (piskvorky) [inline gensim/corpora/indexedcorpus.py]: `numpy.int` and `int` are the same thing; perhaps you meant `numpy.integer`? (to also cover int32, uint16 etc etc).

**Comment by piskvorky** [comment]
I guess we'll have to fix the scipy fblas thingy from #382 first, otherwise the Travis tests just stall :(

**Comment by mattilyra** [comment]
oh snap - I started testing the `scipy/word2vec_inner` changes yesterday but ran into issues installing `scipy-0.10.1`

**Comment by piskvorky** [comment]
It's OK, I've fixed the Travis version to 0.15.1 just now, in commit 1cb6919.

So Travis should now work OK -- if you can merge `develop` into your branch.

**Comment by mattilyra** [comment]
it seems there were some failing tests in https://github.com/piskvorky/gensim/commit/1cb6919b69a0c4bfc178c7b1e9a88e6bff6264a8

**Comment by piskvorky** [comment]
The fails seem unrelated (doc2vec, #386, cc @gojomo ).

This PR seems good to go! Thanks a lot Matti :) Merging.

**Comment by piskvorky** [comment]
Also, what is your preferred CHANGELOG entry for this change (1 short phrase / sentence)?

**Comment by mattilyra** [comment]
Added support for NumPy style fancy indexing to corpus objects.

**Comment by piskvorky** [comment]
Added in 591555e3.

**Pull request merged**